sql: ensure lease expiration is monotonically increasing

vivekmenezes opened the pull request:
This is a fix for both a correctness issue and a performance problem.
1. Correctness issue during schema changes: A node could hand an
expiration time from a lease to a transaction T, and if the
node reacquired a lease for the table at an expiration less than
the existing lease's expiration it could release the existing
lease (with a greater expiration) before the transaction T committed,
thereby invalidating T's deadline.
2. Performance issue: After a lease acquisition if the expiration
time of the new lease was less than the existing lease, it would replace
the table in the cache with the new lease and release the older
lease. But unfortunately it would also not update the table in the
table name cache (the table name cache always uses the lease with
the highest expiration). The table in the table name cache would have
it's lease released rendering the name entry invalid. Therefore until
the node acquired another lease for the same table, the node would be
in a state where there was a table cached but it's name -> table cache
entry invalid, resulting in it making name resolution requests for
every query (but not lease renewals).

It is likely this bug started happening with the addition of more
aggressive lease acquisitions through #28725. I've reduced the value
of DefaultTableDescriptorLeaseJitterFraction to make the logic
ensuring the monotonicity of the lease expiration fire infrequently.

fixes #36044

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36531)
<!-- Reviewable:end -->

andreimatei reviewed (commented):
How can we acquire a lease with a lower expiration time than one we already have? Is that because of the jitter we add to lease durations?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-:-Lbd5FkR4ZLucoxQTepQ:b-hksjd1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
yes, the jitter to the lease duration. I explored potentially removing that but felt this was the better approach.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-:-Lbd9iwo1ME59x2sjbfG:b-crhp0m)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
another option to this implementation is passing a minExpiration timestamp to LeaseStore.acquire()

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-:-LbdCTci9APbvZ2V6jpq:b-dswgq8)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: ensure lease expiration is monotonically increasing

This is a fix for both a correctness issue and a performance problem.
1. Correctness issue during schema changes: A node could hand an
expiration time from a lease to a transaction T, and if the
node reacquired a lease for the table at an expiration less than
the existing lease's expiration it could release the existing
lease (with a greater expiration) before the transaction T committed,
thereby invalidating T's deadline.
2. Performance issue: After a lease acquisition if the expiration
time of the new lease was less than the existing lease, it would replace
the table in the cache with the new lease and release the older
lease. But unfortunately it would also not update the table in the
table name cache (the table name cache always uses the lease with
the highest expiration). The table in the table name cache would have
it's lease released rendering the name entry invalid. Therefore until
the node acquired another lease for the same table, the node would be
in a state where there was a table cached but it's name -> table cache
entry invalid, resulting in it making name resolution requests for
every query (but not lease renewals).

It is likely this bug started happening with the addition of more
aggressive lease acquisitions through #28725. I've reduced the value
of DefaultTableDescriptorLeaseJitterFraction to make the logic
ensuring the monotonicity of the lease expiration fire infrequently.

Release note: None

vivekmenezes commented:
@andreimatei ready for review. Thanks!

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @vivekmenezes)

___
*[pkg/sql/lease.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-Lbj9NcW-TshdUlHmpm8:-Lbj9NcX4YqyTBQX1R3m:b-gu20t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L167)):*
> ```Go
> // being dropped, the error will be errTableDropped.
> func (s LeaseStore) acquire(
> 	ctx context.Context, lastExpiration hlc.Timestamp, tableID sqlbase.ID,
> ```

pls comment the new arg. And call it... minExpiration?
___
*[pkg/sql/lease.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbjAYef5-IH6gP34xcG:-LbjAYef5-IH6gP34xcH:b-kfq5k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L177)):*
> ```Go
> 			// use an expiration based on the lastExpiration to guarantee
> 			// a monotonically increasing expiration.
> 			expiration = lastExpiration.Add(int64(time.Millisecond), 0)
> ```

This doesn't really make sense to me... What's the value of a lease that expires on millisecond after a lease we already have? 
I think that if there's enough time on the current lease left such that a new lease acquisition might give you a lease that's expiring even sooner (or expiring later, but not later enough), what's the point of acquiring that lease? I think the caller should make that determination, and just not call `acquire()`. Is there any difficulty in doing that?

FWIW, I was against doing that refresh loop the way it was done, without an integration with the other background lease requisition processes that we had before. Exactly for reasons like this - we now have two policies for reacquiring leases, one presumably sane, the other apparently less so. But now that we have that loop, I see it as its job to not do crazy things like requesting meaningless leases.


<!-- Sent from Reviewable.io -->

vivekmenezes commented:
The only reason I've structured it the way it is because of the need to
acquire new fresh leases (outside of the periodic lease refresh) when a
version change happens. I have thought about changing that to not use the
acquire node lease code path but then we also have to consider that I'd
like to back port this change, and such a change will come with it's own
set of concerns.

I've structured it to be safe and simple so we can confidently backport the
change. Further this new expiration monotonicity code path is rarely taken
with the new adjustments  made to the lease duration jitter constant.

Furthermore I'd like to do minimal work here given that we know we have to
completely rewrite this code to scale it (epoch based leases)

I'm happy to open up an issue for your concerns for a cleaner outcome.

Vivek

On Fri, 5 Apr 2019, 4:34 PM Andrei Matei, <notifications@github.com> wrote:

> *@andreimatei* requested changes on this pull request.
>
> *Reviewable <https://reviewable.io/reviews/cockroachdb/cockroach/36531>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (waiting on
> @vivekmenezes <https://github.com/vivekmenezes>)
>
> ------------------------------
>
> *pkg/sql/lease.go, line 167 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/36531#-Lbj9NcW-TshdUlHmpm8:-Lbj9NcX4YqyTBQX1R3m:b-gu20t2>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L167>):*
>
> // being dropped, the error will be errTableDropped.func (s LeaseStore) acquire(
> 	ctx context.Context, lastExpiration hlc.Timestamp, tableID sqlbase.ID,
>
> pls comment the new arg. And call it... minExpiration?
> ------------------------------
>
> *pkg/sql/lease.go, line 177 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbjAYef5-IH6gP34xcG:-LbjAYef5-IH6gP34xcH:b-kfq5k1>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L177>):*
>
> 			// use an expiration based on the lastExpiration to guarantee
> 			// a monotonically increasing expiration.
> 			expiration = lastExpiration.Add(int64(time.Millisecond), 0)
>
> This doesn't really make sense to me... What's the value of a lease that
> expires on millisecond after a lease we already have?
> I think that if there's enough time on the current lease left such that a
> new lease acquisition might give you a lease that's expiring even sooner
> (or expiring later, but not later enough), what's the point of acquiring
> that lease? I think the caller should make that determination, and just not
> call acquire(). Is there any difficulty in doing that?
>
> FWIW, I was against doing that refresh loop the way it was done, without
> an integration with the other background lease requisition processes that
> we had before. Exactly for reasons like this - we now have two policies for
> reacquiring leases, one presumably sane, the other apparently less so. But
> now that we have that loop, I see it as its job to not do crazy things like
> requesting meaningless leases.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/36531#pullrequestreview-223467727>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBLRCgNAWssKXNcnjQjAWEOrLIph1ks5vd7NogaJpZM4cdHSu>
> .
>

andreimatei commented:
Hmm so when a version is incremented, it's actually beneficial for a node to get a lease on the new version with the same expiration as the lease on the lease on the old version? Why exactly is that? Is it because having the lease on the new version makes the node release the lease on the old version sooner somehow? Or is it that sometimes a node really needs a lease on the newest version regardless of what other leases it has?
Please put comments about this around that new argument the patch is adding.

But now that you mention version changes, I'm confused. The first reason you list for this patch is the "correctness issue during schema changes" one. But is that addressed in the case where you have a lease on v1, and then you get a lease on v2 and release the one on v1.  You say that it was bad that we were releasing the v1 lease because its expiration has been used for a txn deadline. But don't we still have this problem if we release v1 after we've gotten the lease on v2? Yes, the node still has a lease covering the txn's deadline, but it's a lease on the wrong version. No?

vivekmenezes commented:
I suppose when we discussed this we considered the model where the
acquisition wouldn't happen if the older expiration was sufficient, but
then I saw that we still have to accept leases to pick up new versions, the
forced lease case. So I went with the model where this function always
succeeds by picking a greater time.

Vivek

On Sat, 6 Apr 2019, 2:11 PM Andrei Matei, <notifications@github.com> wrote:

> Hmm so when a version is incremented, it's actually beneficial for a node
> to get a lease on the new version with the same expiration as the lease on
> the lease on the old version? Why exactly is that? Is it because having the
> lease on the new version makes the node release the lease on the old
> version sooner somehow? Or is it that sometimes a node really needs a lease
> on the newest version regardless of what other leases it has?
> Please put comments about this around that new argument the patch is
> adding.
>
> But now that you mention version changes, I'm confused. The first reason
> you list for this patch is the "correctness issue during schema changes"
> one. But is that addressed in the case where you have a lease on v1, and
> then you get a lease on v2 and release the one on v1. You say that it was
> bad that we were releasing the v1 lease because its expiration has been
> used for a txn deadline. But don't we still have this problem if we release
> v1 after we've gotten the lease on v2? Yes, the node still has a lease
> covering the txn's deadline, but it's a lease on the wrong version. No?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/36531#issuecomment-480525452>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBHagQ6QMqp4Kh_7HJOxOV5WtEoztks5veONHgaJpZM4cdHSu>
> .
>

vivekmenezes commented:
The other problem you mention is not a problem because we call upsert which
only replaces a lease if the version is the same

On Sat, 6 Apr 2019, 2:37 PM Vivek Menezes, <vivek@cockroachlabs.com> wrote:

> I suppose when we discussed this we considered the model where the
> acquisition wouldn't happen if the older expiration was sufficient, but
> then I saw that we still have to accept leases to pick up new versions, the
> forced lease case. So I went with the model where this function always
> succeeds by picking a greater time.
>
> Vivek
>
> On Sat, 6 Apr 2019, 2:11 PM Andrei Matei, <notifications@github.com>
> wrote:
>
>> Hmm so when a version is incremented, it's actually beneficial for a node
>> to get a lease on the new version with the same expiration as the lease on
>> the lease on the old version? Why exactly is that? Is it because having the
>> lease on the new version makes the node release the lease on the old
>> version sooner somehow? Or is it that sometimes a node really needs a lease
>> on the newest version regardless of what other leases it has?
>> Please put comments about this around that new argument the patch is
>> adding.
>>
>> But now that you mention version changes, I'm confused. The first reason
>> you list for this patch is the "correctness issue during schema changes"
>> one. But is that addressed in the case where you have a lease on v1, and
>> then you get a lease on v2 and release the one on v1. You say that it was
>> bad that we were releasing the v1 lease because its expiration has been
>> used for a txn deadline. But don't we still have this problem if we release
>> v1 after we've gotten the lease on v2? Yes, the node still has a lease
>> covering the txn's deadline, but it's a lease on the wrong version. No?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cockroachdb/cockroach/pull/36531#issuecomment-480525452>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ALOpBHagQ6QMqp4Kh_7HJOxOV5WtEoztks5veONHgaJpZM4cdHSu>
>> .
>>
>

vivekmenezes commented:
I'm cool with also going with the prior approach where upsert rejects the
new lease if it's expiration time is insufficient. The rejection is very
rare but upsert rejecting it is easier to read IMHO


On Sat, 6 Apr 2019, 2:38 PM Vivek Menezes, <vivek@cockroachlabs.com> wrote:

> The other problem you mention is not a problem because we call upsert
> which only replaces a lease if the version is the same
>
> On Sat, 6 Apr 2019, 2:37 PM Vivek Menezes, <vivek@cockroachlabs.com>
> wrote:
>
>> I suppose when we discussed this we considered the model where the
>> acquisition wouldn't happen if the older expiration was sufficient, but
>> then I saw that we still have to accept leases to pick up new versions, the
>> forced lease case. So I went with the model where this function always
>> succeeds by picking a greater time.
>>
>> Vivek
>>
>> On Sat, 6 Apr 2019, 2:11 PM Andrei Matei, <notifications@github.com>
>> wrote:
>>
>>> Hmm so when a version is incremented, it's actually beneficial for a
>>> node to get a lease on the new version with the same expiration as the
>>> lease on the lease on the old version? Why exactly is that? Is it because
>>> having the lease on the new version makes the node release the lease on the
>>> old version sooner somehow? Or is it that sometimes a node really needs a
>>> lease on the newest version regardless of what other leases it has?
>>> Please put comments about this around that new argument the patch is
>>> adding.
>>>
>>> But now that you mention version changes, I'm confused. The first reason
>>> you list for this patch is the "correctness issue during schema changes"
>>> one. But is that addressed in the case where you have a lease on v1, and
>>> then you get a lease on v2 and release the one on v1. You say that it was
>>> bad that we were releasing the v1 lease because its expiration has been
>>> used for a txn deadline. But don't we still have this problem if we release
>>> v1 after we've gotten the lease on v2? Yes, the node still has a lease
>>> covering the txn's deadline, but it's a lease on the wrong version. No?
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/cockroachdb/cockroach/pull/36531#issuecomment-480525452>,
>>> or mute the thread
>>> <https://github.com/notifications/unsubscribe-auth/ALOpBHagQ6QMqp4Kh_7HJOxOV5WtEoztks5veONHgaJpZM4cdHSu>
>>> .
>>>
>>

andreimatei commented:
I have to read a bunch of code to have an informed opinion here.
If you're in a here, LGTM on the current version, but please put some
comments.

On Sat, Apr 6, 2019, 3:42 PM vivekmenezes <notifications@github.com> wrote:

> I'm cool with also going with the prior approach where upsert rejects the
> new lease if it's expiration time is insufficient. The rejection is very
> rare but upsert rejecting it is easier to read IMHO
>
>
> On Sat, 6 Apr 2019, 2:38 PM Vivek Menezes, <vivek@cockroachlabs.com>
> wrote:
>
> > The other problem you mention is not a problem because we call upsert
> > which only replaces a lease if the version is the same
> >
> > On Sat, 6 Apr 2019, 2:37 PM Vivek Menezes, <vivek@cockroachlabs.com>
> > wrote:
> >
> >> I suppose when we discussed this we considered the model where the
> >> acquisition wouldn't happen if the older expiration was sufficient, but
> >> then I saw that we still have to accept leases to pick up new versions,
> the
> >> forced lease case. So I went with the model where this function always
> >> succeeds by picking a greater time.
> >>
> >> Vivek
> >>
> >> On Sat, 6 Apr 2019, 2:11 PM Andrei Matei, <notifications@github.com>
> >> wrote:
> >>
> >>> Hmm so when a version is incremented, it's actually beneficial for a
> >>> node to get a lease on the new version with the same expiration as the
> >>> lease on the lease on the old version? Why exactly is that? Is it
> because
> >>> having the lease on the new version makes the node release the lease
> on the
> >>> old version sooner somehow? Or is it that sometimes a node really
> needs a
> >>> lease on the newest version regardless of what other leases it has?
> >>> Please put comments about this around that new argument the patch is
> >>> adding.
> >>>
> >>> But now that you mention version changes, I'm confused. The first
> reason
> >>> you list for this patch is the "correctness issue during schema
> changes"
> >>> one. But is that addressed in the case where you have a lease on v1,
> and
> >>> then you get a lease on v2 and release the one on v1. You say that it
> was
> >>> bad that we were releasing the v1 lease because its expiration has been
> >>> used for a txn deadline. But don't we still have this problem if we
> release
> >>> v1 after we've gotten the lease on v2? Yes, the node still has a lease
> >>> covering the txn's deadline, but it's a lease on the wrong version. No?
> >>>
> >>> —
> >>> You are receiving this because you were mentioned.
> >>> Reply to this email directly, view it on GitHub
> >>> <
> https://github.com/cockroachdb/cockroach/pull/36531#issuecomment-480525452
> >,
> >>> or mute the thread
> >>> <
> https://github.com/notifications/unsubscribe-auth/ALOpBHagQ6QMqp4Kh_7HJOxOV5WtEoztks5veONHgaJpZM4cdHSu
> >
> >>> .
> >>>
> >>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/36531#issuecomment-480532045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBce_uX3_ToHLpJyRZP5RaTLHfoZQIks5vePiegaJpZM4cdHSu>
> .
>

vivekmenezes commented:
@andreimatei I've added a function hasValidExpiration() and added some more comments to make the code clearer. upsertLocked() also returns an error when it violates the expiration time monotonicity, which can actually never happen, but I suppose we are not allowed to panic any more :-)

Thanks for taking a look.

andreimatei reviewed (approved):
I don't want to block merging this if it's important and you're happy with the patch. So feel free to merge as far as I'm concerned.

But I'm looking at it over and over and it still doesn't really make sense to me. I have a feeling that something else is the answer, but I also don't know the code well enough any more to be absolutely positive. Fundamentally, it seems to me that we're acquiring useless leases and then we're adding code so that these useless leases don't end up being to disturbing. I still think that we shouldn't be acquiring leases at all. You've suggested something about how sometimes we have to acquire leases regardless of the expiration of existing ones when we find out about new tables versions, and I bought that for a while but now that I'm thinking about it, I don't know if I buy it any more... If we're acquiring a lease on a new version, I think it should be fine if its expiration is lower than a lease on an older version, no? How do leases on different versions interact with each other exactly?

The path cites two reasons for its existence:
1. "Correctness issue during schema changes...  if the
node reacquired a lease for the table at an expiration less than
the existing lease's expiration it could release the existing
lease (with a greater expiration) before the transaction T committed,
thereby invalidating T's deadline." - well so why exactly are we releasing the old lease if the newest one we got is worse? And again, why exactly are we getting a new one to begin with?

2. "Performance issue... if the expiration
time of the new lease was less than the existing lease, it would replace
the table in the cache with the new lease and release the older
lease" - same as above - can't the new lease not replace the old one, and so the name cache would continue to refer to a valid lease?

> I'm cool with also going with the prior approach where upsert rejects the new lease if it's expiration time is insufficient. 

Yeah, maybe that is better. I don't remember what the patch was doing originally, but perhaps it was indeed better than the current version. My objection to it was that we were acquiring a seemingly useless lease, right? That objection remains with the current version too, so I guess if we can't do without this funky acquisition, perhaps I'd rather chose the old version :P. But I'm still skeptical that we need to acquire these leases with low expirations in the first place. Maybe you could spell that out for me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-:-LbxUhWfAWTPjv2KgIOJ:b8ea4f3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @vivekmenezes)

___
*[pkg/sql/lease.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxS_P7Ef9dBikvyf6z:-LbxS_P7Ef9dBikvyf7-:bjf5sy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L109)):*
> ```Go
> }
> 
> // Does this table have a larger expiration than the existing
> ```

nit: start the comment with `hasValidExpiration ...`, like for exported methods.
___
*[pkg/sql/lease.go, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxSgFmBP4p0FED6f9t:-LbxSgFmBP4p0FED6f9u:b1rppo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L113)):*
> ```Go
> // of the expiration times on a table at a particular version.
> // The version is not explicitly checked here.
> func (s *tableVersionState) hasValidExpiration(existing *tableVersionState) bool {
> ```

I for one would do away with this function. What's a "valid expiration"? What's the "monotonicity of the expirations times"? I think just inlining it is better.
___
*[pkg/sql/lease.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxU9yqCYVcnyRs4ybN:-LbxU9yqCYVcnyRs4ybO:bdh9jbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L175)):*
> ```Go
> // being dropped, the error will be errTableDropped.
> func (s LeaseStore) acquire(
> 	ctx context.Context, minExpiration hlc.Timestamp, tableID sqlbase.ID,
> ```

pls put a comment on the new argument
___
*[pkg/sql/lease.go, line 895 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxQ3AT07GX_caNZuLZ:-LbxQ3AU0M5SUCFMnG4p:b-1abj6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L895)):*
> ```Go
> 		// This is a violation of an invariant and can actually not
> 		// happen. We return an error here to aid in further investigations.
> 		return nil, errors.Errorf("lease expiration monotonicity violation, (%+v) vs (%+v)", s, table)
> ```

nit: `%+v`? The `String()` method on `storedTableLease` looks pretty good to me...
___
*[pkg/sql/lease.go, line 944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxRD2h2JWvesPAHPjI:-LbxRD2h2JWvesPAHPjJ:bn8305o) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L944)):*
> ```Go
> // If the lease cannot be obtained because the descriptor is in the process of
> // being dropped, the error will be errTableDropped.
> // minExpirationTime, if not set to the zero value, will be used as a lower
> ```

pls remove this comment on `minExpirationTime` while you're here


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/sql/lease.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-Lbj9NcW-TshdUlHmpm8:-LbxrMKGDolwjNXm5c9v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L167)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls comment the new arg. And call it... minExpiration?
</blockquote></details>

Done.
___
*[pkg/sql/lease.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbjAYef5-IH6gP34xcG:-LbyGjZi7S85kB86SNFh:b-xg67oj) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L177)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

This doesn't really make sense to me... What's the value of a lease that expires on millisecond after a lease we already have? 
I think that if there's enough time on the current lease left such that a new lease acquisition might give you a lease that's expiring even sooner (or expiring later, but not later enough), what's the point of acquiring that lease? I think the caller should make that determination, and just not call `acquire()`. Is there any difficulty in doing that?

FWIW, I was against doing that refresh loop the way it was done, without an integration with the other background lease requisition processes that we had before. Exactly for reasons like this - we now have two policies for reacquiring leases, one presumably sane, the other apparently less so. But now that we have that loop, I see it as its job to not do crazy things like requesting meaningless leases.
</blockquote></details>

the reason why we acquire a lease is because we don't know whether we're acquiring for the current or a new version. And we don't know whether we have a new version unless we've read the descriptor from the store as part of the lease acquisition.
___
*[pkg/sql/lease.go, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxSgFmBP4p0FED6f9t:-LbxsQMPDCZRqcj8Gc2G:b-yr6tzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L113)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I for one would do away with this function. What's a "valid expiration"? What's the "monotonicity of the expirations times"? I think just inlining it is better.
</blockquote></details>

I've added it so that the table descriptor cache and the name cache use the same function.
___
*[pkg/sql/lease.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxU9yqCYVcnyRs4ybN:-LbxrVWL3Id_05lgQts2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L175)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls put a comment on the new argument
</blockquote></details>

Done.
___
*[pkg/sql/lease.go, line 944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbxRD2h2JWvesPAHPjI:-LbxryWNB5trdLn1H9mR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7bf0cb228c905f2a8cabeb91336e2295ee7bf6/pkg/sql/lease.go#L944)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls remove this comment on `minExpirationTime` while you're here
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

vivekmenezes commented:
bors r+

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @vivekmenezes)

___
*[pkg/sql/lease.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbjAYef5-IH6gP34xcG:-LbyM42s7qHbOBBmcKlx:b-w1tytc) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L177)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

the reason why we acquire a lease is because we don't know whether we're acquiring for the current or a new version. And we don't know whether we have a new version unless we've read the descriptor from the store as part of the lease acquisition.
</blockquote></details>

but when we're initiating a lease acquisition because we've find out that there's a new version around, we do know, right? So those cases could do their lease acquisition and be certain they're getting a new version, whereas other could be more discriminate and not acquire when it's not certain that they'll get a lease with a higher expiration time than what we have already, I think.


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36531)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/sql/lease.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36531#-LbjAYef5-IH6gP34xcG:-LbyR8o6DoW1d5Vsr1H_:bj2xqbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a257aa42ed2ad5ebce2a626bcc59ea04428319/pkg/sql/lease.go#L177)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

but when we're initiating a lease acquisition because we've find out that there's a new version around, we do know, right? So those cases could do their lease acquisition and be certain they're getting a new version, whereas other could be more discriminate and not acquire when it's not certain that they'll get a lease with a higher expiration time than what we have already, I think.
</blockquote></details>

Given the way the code is structured AcquireFreshestFromStore() is called in two places. I actually have another PR that get rid of one of them so perhaps then I could consider making this more explicitly not acquire a lease. But for 19.1 that is all very risky and I don't want to pursue that. I do worry we'll hit a race where we don't acquire a lease when we ought to.


<!-- Sent from Reviewable.io -->

vivekmenezes commented:
bors r+

vivekmenezes commented:
bors r+
